clang/include/clang/Sema/BoundsWideningAnalysis.h

```diff
@@ -56,14 +67,139 @@ namespace clang {
+    void ComputeKillSet(ElevatedCFGBlock *EB, const Stmt *S,
+                        const Stmt *PrevS);
+
+    // Get all variables modfied by statement S or statements nested in S.
+    // @param[in] S is a given statement.
+    // @param[out] ModifiedVars is a set of variables modified by S or
+    // statements nested in S.
+    void GetModifiedVars(const Stmt *S, VarSetTy &ModifiedVars);
+
+    // Add an offset to a given expression to get the widened expression.
+    // @param[in] E is the given expression.
+    // @param[in] Offset is the given offset.
+    // @return Returns the expression E + Offset.
+    Expr *GetWidenedExpr(Expr *E, unsigned Offset) const;
+
+    // From a given terminating condition extract the terminating condition for
+    // the current block. Given an expression like "if (e1 && e2)" this
+    // function returns e2 which is the terminating condition for the current
+    // block.
+    // @param[in] E is given terminating condition.
+    // @return The terminating condition for the block.
+    Expr *GetTerminatorCondition(const Expr *E) const;
+
+    // Use the last statement in a block to get the terminating condition for
+    // the block. This could be an expression of the form "if (e1 && e2)".
+    // @param[in] B is the block for which we need the terminating condition.
+    // @return Expression for the terminating condition of block B.
+    Expr *GetTerminatorCondition(const CFGBlock *B) const;
+
+    // From the given expression get the dereferences expression. A dereference
+    // expression can be of the form "*(p + 1)" or "p[1]".
+    // @param[in] E is the given expression.
+    // return Returns the dereference expression, if it exists.
+    Expr *GetDerefExpr(Expr *E) const;
+
+    // Get the variables occurring in an expression.
+    // @param[in] E is the given expression.
+    // @param[out] VarsInExpr is a set of variables that occur in E.
+    void GetVarsInExpr(Expr *E, VarSetTy &VarsInExpr) const;
+
+    // Invoke IgnoreValuePreservingOperations to strip off casts.
+    // @param[in] E is the expression whose casts must be stripped.
+    // @return E with casts stripped off.
+    Expr *IgnoreCasts(const Expr *E) const;
+
+    // We do not want to run dataflow analysis on null blocks or the exit
+    // block. So we skip them.
+    // @param[in] B is the block which may need to the skipped from dataflow
+    // analysis.
+    // @return Whether B should be skipped.
+    bool SkipBlock(const CFGBlock *B) const;
+
+    // Check if V is an _Nt_array_ptr or an _Nt_checked array.
+    // @param[in] V is a VarDecl.
+    // @return Whether V is an _Nt_array_ptr or an _Nt_checked array.
+    bool IsNtArrayType(const VarDecl *V) const;
+
+    // Compute the set difference of sets A and B.
+    // @param[in] A is a set.
+    // @param[in] B is a set.
+    // @return The set difference of sets A and B.
+    template<class T, class U> T Difference(T &A, U &B) const;
```

Comment:
Would it make sense to put `Difference` and `Set` in ExprUtils.h or another utility file?